### PR TITLE
[Core] Add variables to registry

### PR DESCRIPTION
**Pull request by pooyan-dadvand**
**📝 Description**
This is a continuation of #9308 and should be merged after that

Here I am adding the automatic registration directly to the variables constructors. This is assuming that we don't construct variables in the performance-intensive parts. 

The copy constructor doesn't register because it assumes that the variable should have been already registered.

The registry will keep a copy of that variable so we can use this copy in our data value container to avoid problems with local variables that may go out of scope:  #3789

**🆕 Changelog**
- Adding RegisterVariables to constructor of the Variables

### CODE REVIEW

**Comment by pooyan-dadvand** [comment]
I think it is better to add this PR as it is and add the type in a separate branch.

 @KratosMultiphysics/technical-committee would you please review this?

File: kratos/includes/registry_value_item.h [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+/// The registry value item is a registry item which stores a copy of a value.
+/** This class is derived from RegistryItem and the difference is that it creates a 
+ *  copy of the value in initialization and delete that copy at destruction. 
+ *  This make it suitable for storing values that we want to have one copy of them 
+ *  in the registry but not for the values that are stored in other places (like static variables) 
+ *  Value type should be copy constructable.
+*/
+template<typename TValueType>
+class KRATOS_API(KRATOS_CORE) RegistryValueItem : public RegistryItem
```
Comment (roigcarlo) [inline kratos/includes/registry_value_item.h]: This `KRATOS_API` is causing the win problems. It should not be here unless there is a cpp for RegistryValueItem.

**Comment by pooyan-dadvand** [comment]
@sunethwarna would you please check why `test_ShapeOptimizationLiftConstrainedBodyFitted2D` is failing?

Unfortunately, I cannot reproduce the problem in my machine to debug it and I don't have any clue about this solver....

File: kratos/includes/registry_value_item.h (hunk 1 above)
Comment (pooyan-dadvand) [reply kratos/includes/registry_value_item.h]: You are right! Thanks for the hint!

**Comment by pooyan-dadvand** [comment]
@loumalouomega while executing all tests in FullDebug I found out the following error in `ALMStaticEvolutionLoadFrictionTestContact)`:

```
RuntimeError: Error: The following errors occured in a parallel region!
Thread #25 caught exception: Error: attempting to do a GetValue for: AUGMENTED_NORMAL_CONTACT_PRESSURE variable #2460668928AUGMENTED_NORMAL_CONTACT_PRESSURE variable #2460668928 #2460668928 unfortunately the variable is not in the database and the operations is not threadsafe (this function is being called from within a parallel region)
```

**Comment by loumalouomega** [comment]
> @loumalouomega while executing all tests in FullDebug I found out the following error in `ALMStaticEvolutionLoadFrictionTestContact)`:
> 
> ```
> RuntimeError: Error: The following errors occured in a parallel region!
> Thread #25 caught exception: Error: attempting to do a GetValue for: AUGMENTED_NORMAL_CONTACT_PRESSURE variable #2460668928AUGMENTED_NORMAL_CONTACT_PRESSURE variable #2460668928 #2460668928 unfortunately the variable is not in the database and the operations is not threadsafe (this function is being called from within a parallel region)
> ```

Thanks, i will take a look (a priori this variable is initialized at the beginning)

**Comment by loumalouomega** [comment]
> > @loumalouomega while executing all tests in FullDebug I found out the following error in `ALMStaticEvolutionLoadFrictionTestContact)`:
> > ```
> > RuntimeError: Error: The following errors occured in a parallel region!
> > Thread #25 caught exception: Error: attempting to do a GetValue for: AUGMENTED_NORMAL_CONTACT_PRESSURE variable #2460668928AUGMENTED_NORMAL_CONTACT_PRESSURE variable #2460668928 #2460668928 unfortunately the variable is not in the database and the operations is not threadsafe (this function is being called from within a parallel region)
> > ```
> 
> Thanks, I will take a look (a priori this variable is initialized at the beginning)

**Comment by pooyan-dadvand** [comment]
In the DEM test: 
```
Error: Tests for DEMApplication took too long. Process Killed.
```

@KratosMultiphysics/dem  would you please check if you can reduce the test time consumption?

**Comment by farrufat-cimne** [comment]
> In the DEM test:
> 
> ```
> Error: Tests for DEMApplication took too long. Process Killed.
> ```
> 
> @KratosMultiphysics/dem would you please check if you can reduce the test time consumption?

Should be fixed in #9370.

**Comment by loumalouomega** [comment]
> > @loumalouomega while executing all tests in FullDebug I found out the following error in `ALMStaticEvolutionLoadFrictionTestContact)`:
> > ```
> > RuntimeError: Error: The following errors occured in a parallel region!
> > Thread #25 caught exception: Error: attempting to do a GetValue for: AUGMENTED_NORMAL_CONTACT_PRESSURE variable #2460668928AUGMENTED_NORMAL_CONTACT_PRESSURE variable #2460668928 #2460668928 unfortunately the variable is not in the database and the operations is not threadsafe (this function is being called from within a parallel region)
> > ```
> 
> Thanks, I will take a look (a priori this variable is initialized at the beginning)

#9372

**Comment by rubenzorrilla** [comment]
In this regard @KratosMultiphysics/technical-committee agreed to
1. Merge @sunethwarna 's registry enhancements
2. Merge tha master in here to check if these whether fix the problem or not

**Comment by sunethwarna** [comment]
Recent merge fixed the windows problem. So I think this can go in.... @pooyan-dadvand @rubenzorrilla @roigcarlo

**Review by rubenzorrilla — APPROVED** [review]
Approving on behalf of the KTC as the problem is already solved.

**Pull request merged**